### ISSUE TITLE
ApolloClient 3: Cannot update a component from inside the function body of a different component.

### ISSUE THREAD

**Issue by Banou26** [apollographql/apollo-client#6188]
*related to https://github.com/apollographql/react-apollo/issues/3863*

**Intended outcome:**
`useQuery` should not throw unintentional state changes warning.

**Actual outcome:**

Error thrown:
`Warning: Cannot update a component from inside the function body of a different component.`

**How to reproduce the issue:**
probably the same as *https://github.com/apollographql/react-apollo/issues/3863#issuecomment-608573622*

**Versions**
`"@apollo/client": "^3.0.0-beta.44"`

**Comment by Tadimsky** [comment]
This is similar to the error I get - I believe it's just a different string in React 16.13.1:

```
Warning: Cannot update a component (`Participants`) while rendering a different component (`HandRaiseItem`). To locate the bad setState() call inside `HandRaiseItem`, follow the stack trace as described in https://fb.me/setstate-in-render
```

@hwillson any chance you could fix this here as well?

**Comment by hwillson** [comment]
Is this still an issue for people using the latest `@apollo/client@beta` (3.0.0-beta.50)?

**Comment by Banou26** [comment]
Currently on `^3.0.0-beta.48` and i don't have it anymore.

**Comment by fdev** [comment]
This issue was indeed fixed for me in `beta.48`, but seems to have returned in `beta.52`. @Banou26 can you check if the latest version produces the error for you again as well?

**Comment by Banou26** [comment]
Tried on `^3.0.0-beta.53` and it's not back.

**Comment by fdev** [comment]
For some reason `npm install '@apollo/client@3.0.0-beta.52'` kept installing version `3.0.0-james.33.0`, which is why it was broken again for me. This no longer happens with `3.0.0-beta.53`. Thanks for checking!

**Comment by tpoppen** [comment]
I am experiencing this issue on `@apollo/client@3.0.0-rc.10`

**Comment by coler-j** [comment]
I am getting this when using `reactiveVars` for pagination in a child component.

**Comment by AdrienLemaire** [comment]
@coler-j I've just encountered the same issue, and fixed it by adding the reactive var mutation in an effect.

In your case, that would be:
```ts
export default function App() {
  const { data } = useQuery(. .......... )
  useEffect(() => {
    currentDataVar(data);
  }, [data]);

  return (<>
     <SomeComponentToRenderData />  
     <SomeChildComponent />
   </>
  )
}
```

**Comment by asaadawey** [comment]
@coler-j 
Same issue of you
did you got any solution ?

**Comment by coler-j** [comment]
@asaadawey I have not circled back but I anticipate that this https://github.com/apollographql/apollo-client/issues/6188#issuecomment-796382651 comment should fix it. Although this is not in the docs, effects that impact state should typically be wrapped in `useEffect`

**Comment by asaadawey** [comment]
@coler-j  
Thank you for replaying
actually i have got another solution
warpping the code with ```setTimeout```

**Comment by Banou26** [comment]
Should i re-open this issue if it's still a problem?

**Comment by asaadawey** [comment]
@Banou26 
i think yes

wrapping the apollo code inside a ```useEffect``` or ```setTimeout``` is not an efficient solution

**Comment by alexistbell** [comment]
Any update on this issue? Since I started using pagination I have run into this problem. I tried reworking to put the code in side of useEffect instead, but it doesn't seem to make a difference.

**Comment by dylanwulf** [comment]
@alexistbell I'm not sure if it's exactly the same issue, but I submitted this PR which might fix it: #9801. This change is available in version `3.7.0-beta.6` and newer.

**Comment by alexistbell** [comment]
Thanks @dylanwulf , that is making the error go away, which is helpful because it looks like I have a different issue I need to track down that isn't related, so now I can focus on that.
